File: sdk/python/nova_core_sdk/hooks.py
from __future__ import annotations

from typing import Any, Callable


class HookManager:
    def __init__(self) -> None:
        self._hooks: dict[str, list[Callable[..., Any]]] = {
            "pre_request": [],
            "post_request": [],
            "pre_auth": [],
            "post_auth": [],
            "on_error": [],
            "on_retry": [],
            "on_stream_chunk": [],
        }

    def register(self, event: str, callback: Callable[..., Any]) -> None:
        if event not in self._hooks:
            self._hooks[event] = []
        self._hooks[event].append(callback)

    def unregister(self, event: str, callback: Callable[..., Any]) -> None:
        if event in self._hooks and callback in self._hooks[event]:
            self._hooks[event].remove(callback)

    def emit(self, event: str, *args: Any, **kwargs: Any) -> list[Any]:
        results: list[Any] = []
        for callback in self._hooks.get(event, []):
            result = callback(*args, **kwargs)
            results.append(result)
        return results

    def clear(self, event: str | None = None) -> None:
        if event:
            self._hooks.pop(event, None)
        else:
            self._hooks.clear()

File: sdk/typescript/src/hooks.ts
type HookCallback = (...args: unknown[]) => unknown;

export class HookManager {
  private hooks: Map<string, Set<HookCallback>> = new Map();

  register(event: string, callback: HookCallback): void {
    if (!this.hooks.has(event)) {
      this.hooks.set(event, new Set());
    }
    this.hooks.get(event)!.add(callback);
  }

  unregister(event: string, callback: HookCallback): void {
    this.hooks.get(event)?.delete(callback);
  }

  emit(event: string, ...args: unknown[]): unknown[] {
    const results: unknown[] = [];
    this.hooks.get(event)?.forEach((callback) => {
      results.push(callback(...args));
    });
    return results;
  }

  clear(event?: string): void {
    if (event) {
      this.hooks.delete(event);
    } else {
      this.hooks.clear();
    }
  }
}
